ddl/table.go

```diff
@@ -948,56 +956,191 @@ func updateVersionAndTableInfo(t *meta.Meta, job *model.Job, tblInfo *model.Tabl
+		if partitionDefInMeta != nil {
+			if partitionDefInMeta.State == model.StatePublic {
+				// Even one partition with the same name existed, whole job canceled.
+				// We already have these new added partition with the same name.
+				job.State = model.JobStateCancelled
+				return nil, nil, nil, errors.Trace(ErrSameNamePartition.GenWithStackByArgs(pd.Name.L))
+			}
+			// Collect the partition definition pointer in tableInfo for future state change.
+			partDefPointers = append(partDefPointers, partitionDefInMeta)
+		}
 	}
+	return tblInfo, partInfo, partDefPointers, nil
+}
 
-	tblInfo, err := getTableInfoAndCancelFaultJob(t, job, job.SchemaID)
+func checkPartitionReplica(partDefPointers []*model.PartitionDefinition, d *ddlCtx, job *model.Job) (needWait bool, err error) {
+	ctx := context.Background()
+	pdCli := d.store.(tikv.Storage).GetRegionCache().PDClient()
+	stores, err := pdCli.GetAllStores(ctx)
 	if err != nil {
 		job.State = model.JobStateCancelled
-		return ver, errors.Trace(err)
+		return needWait, errors.Trace(err)
 	}
-	err = checkAddPartitionTooManyPartitions(uint64(len(tblInfo.Partition.Definitions) + len(partInfo.Definitions)))
-	if err != nil {
-		job.State = model.JobStateCancelled
-		return ver, errors.Trace(err)
+	for _, pd := range partDefPointers {
+		startKey, endKey := tablecodec.GetTableHandleKeyRange(pd.ID)
+		regions, _, err := pdCli.ScanRegions(ctx, startKey, endKey, -1)
+		if err != nil {
+			job.State = model.JobStateCancelled
+			return needWait, errors.Trace(err)
+		}
+		// For every region in the partition, if it has some corresponding peers and
+		// no pending peers, that means the replication has completed.
+		for _, region := range regions {
+			regionState, err := pdCli.GetRegionByID(ctx, region.Id)
+			if err != nil {
+				job.State = model.JobStateCancelled
+				return needWait, errors.Trace(err)
+			}
+			tiflashPeerCount := checkTiFlashPeerStore(stores, regionState.Meta.Peers)
+			// It's unnecessary to wait all tiflash peer to be replicated.
+			// Here only make sure that tiflash peer count > 0 (at least one).
+			if tiflashPeerCount > 0 {
+				continue
+			}
+			needWait = true
+			logutil.Logger(ctx).Info("[DDL] partition replica check failed in delete-only ddl state", zap.Int64("pid", pd.ID), zap.Uint64("wait region ID", region.Id), zap.Uint64("tiflash peer nums", tiflashPeerCount), zap.Time("check time", time.Now()))
+			return needWait, nil
+		}
 	}
+	logutil.Logger(ctx).Info("[DDL] partition replica checks ok in delete-only ddl state")
+	return needWait, nil
+}
 
-	err = checkAddPartitionValue(tblInfo, partInfo)
-	if err != nil {
-		job.State = model.JobStateCancelled
-		return ver, errors.Trace(err)
+func checkTiFlashPeerStore(stores []*metapb.Store, peers []*metapb.Peer) (tiFlashPeerCount uint64) {
+	for _, peer := range peers {
+		for _, store := range stores {
+			if peer.StoreId == store.Id && storeHasEngineTiFlashLabel(store) {
+				tiFlashPeerCount++
```

Comment:
We only need to check if one store has tiflash label？Then we can remove the check of https://github.com/pingcap/tidb/pull/18495/files#diff-71209e5c7ffc793d7bf8779373b48098R1014 and reduce traversal.